### ISSUE TITLE
ci: file_diff_from_base does not work for ghstack PRs on GHA

### ISSUE THREAD

**Issue by seemethere** [pytorch/pytorch#60111]
# Problem

The function `file_diff_from_base` does not currently work for ghstack PRs on GHA.

Previous attempts at this can be found in the commits from April 23, 2021 here:
* https://github.com/pytorch/pytorch/pull/56494/commits

With failures mostly looking like:
```
+ file_diff_from_base /tmp/tmp.cgmJJ3LQMK
+ set +e
+ git fetch origin master --quiet
+ set -e
+ [[ -n true ]]
+ git diff --name-only e2fe5924201ed1e7cc48aaa6cb21732c694c34fc d1cae0dae7602df47d9d1981e44116f216960c4a
fatal: bad object e2fe5924201ed1e7cc48aaa6cb21732c694c34fc
```

Logs: https://github.com/pytorch/pytorch/runs/2423051948

_NOTE_: Logs may have disappeared since GHA only keeps logs for ~30days

# Context

The reason why I think this happens is due to the way `ghstack` creates the branches needed to make the stacked PRs work in the first place.

Basically `ghstack` creates references that aren't really real references so you can't find a merge base between the two references, nor create a real diff between the two references.

cc @ezyang @seemethere @malfet @walterddr @pytorch/pytorch-dev-infra

**Comment by ezyang** [comment]
I don't think this is quite right. ghstack's head and base must in fact have a common merge-base, otherwise GitHub's UI wouldn't be able to compute the diff to show. Is the problem that the head and base are not related to master? We briefly broke this but fixed it later. https://github.com/ezyang/ghstack/issues/52

**Comment by seemethere** [comment]
> I don't think this is quite right. ghstack's head and base must in fact have a common merge-base, otherwise GitHub's UI wouldn't be able to compute the diff to show. Is the problem that the head and base are not related to master? We briefly broke this but fixed it later. [ezyang/ghstack#52](https://github.com/ezyang/ghstack/issues/52)

It might also have something to do with the default action of [`actions/checkout`](https://github.com/actions/checkout) not checking out the correct branches in order to find a merge base that works.
